### PR TITLE
nrf52840 support for qspi flash

### PR DESCRIPTION
**Pull request by kasjer**
This adds support for QSPI to NRF52840 MCU and NRF52840PDK

### CODE REVIEW

**Comment by sterlinghughes** [comment]
👍

File: hw/mcu/nordic/nrf52xxx/src/hal_qspi.c [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+#include "os/mynewt.h"
+#if MYNEWT_VAL(QSPI_ENABLE)
+#include <mcu/cmsis_nvic.h>
+#include <hal/hal_flash_int.h>
+#include "mcu/nrf52_hal.h"
+#include "nrf.h"
+#include <nrfx/hal/nrf_qspi.h>
+
+static int
+nrf52k_qspi_read(const struct hal_flash *dev, uint32_t address,
```
Comment (mlaz) [inline hw/mcu/nordic/nrf52xxx/src/hal_qspi.c]: Why nrf52k ?

Comment (kasjer) [reply hw/mcu/nordic/nrf52xxx/src/hal_qspi.c]: same naming convention was used in hw/mcu/nordic/nrf52xxx/src/hal_flash.h
nrf52k_flash_read

**Comment by kasjer** [comment]
@wes3 I though about putting all definition in MCU. I moved some of them to the BSP because there are not sensible default values for pins or flash size. If you think that it is more mynewt way to put them in MCU and then override them in BSP I have no objections.

**Comment by sterlinghughes** [comment]
@kasjer I agree with @wes3, i think they should be defined in the MCU package.  I see your point regarding lack of sensible values -- the BSP will have to override them, but they are specific to configuring the Nordic MCU, and so that's where they should be defined.

**Comment by wes3** [comment]
@kasjer I understand about the lack of sensible defaults. One way I have seen this done (and I am not saying this is good or bad) is that "nonsensical" values are assigned to these values by default. These nonsensical values are then checked in the code and if they are still nonsensical, meaning not overridden somewhere, the code fails to compile. I guess this could be done run time but generally it is done by the preprocessor. For example:

#if MYNEWT_VAL(QPSI_FLASH_SECTOR_SIZE) == 0
#error "Must override qspi flash sector size"
#endif

Or some such thing; I am sure you get the idea.

**Comment by kasjer** [comment]
@wes3 I agree, I will move definitions to MCU and I will add preprocessor checks to detect unchanged values as you suggested, thanks.

**Comment by utzig** [comment]
LGTM

File: hw/mcu/nordic/nrf52xxx/src/hal_qspi.c [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+#include "os/mynewt.h"
+#if MYNEWT_VAL(QSPI_ENABLE)
+#include <mcu/cmsis_nvic.h>
+#include <hal/hal_flash_int.h>
+#include "mcu/nrf52_hal.h"
+#include "nrf.h"
+#include <nrfx/hal/nrf_qspi.h>
+
+#if MYNEWT_VAL(QSPI_FLASH_SECTOR_COUNT) < 1
+#error QSPI_FLASH_SECTOR_COUNT must be set to the correct value in bsp.yml
```
Comment (mkiiskila) [inline hw/mcu/nordic/nrf52xxx/src/hal_qspi.c]: Not in bsp.yml but BSP's syscfg.yml. The same is true for the subsequent #error messages.

File: hw/mcu/nordic/nrf52xxx/src/hal_qspi.c [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    const char src_not_in_ram = (((uint32_t) src) & 0xE0000000) != 0x20000000;
+
+    while ((NRF_QSPI->STATUS & QSPI_STATUS_READY_Msk) == 0)
+        ;
+
+    while (num_bytes != 0) {
+        const uint32_t page_start = address
+                & ~(MYNEWT_VAL(QSPI_FLASH_PAGE_SIZE) - 1);
+        const uint32_t page_limit = page_start
+                + MYNEWT_VAL(QSPI_FLASH_PAGE_SIZE);
```
Comment (mkiiskila) [inline hw/mcu/nordic/nrf52xxx/src/hal_qspi.c]: Coding style: move the page_start, page_limit declaration in the beginning of the function.
When you have a statement which expands to multiple lines, put the operator at the end of the line.
See coding style doc in the root of the repository.

File: hw/mcu/nordic/nrf52xxx/syscfg.yml [hunk 4]
```diff
@@ -114,6 +114,76 @@ syscfg.defs:
         description: 'SOFT PWM'
         value: 0
 
+    QSPI_ENABLE:
+        descritption: 'NRF52 QSPI'
+        value: 0
+
+    QSPI_READOC:
+        descritption: >
```
Comment (mkiiskila) [inline hw/mcu/nordic/nrf52xxx/syscfg.yml]: Typo: descritption should be description. Alos see QSPI_WRITEOC

File: hw/mcu/nordic/nrf52xxx/syscfg.yml [hunk 5]
```diff
@@ -114,6 +114,76 @@ syscfg.defs:
...
+    QSPI_SCK_FREQ:
+        description: '32MHz clock divider (0-31). Clock = 32MHz / (1+divider)'
+        value: 0
+    QSPI_SPI_MODE:
+        description: 'SPI 0=Mode0 or 1=Mode3'
+        value: 0
+
+    QSPI_FLASH_SECTOR_SIZE:
+        description: 'QSPI sector size'
+        value: 4096
```
Comment (mkiiskila) [inline hw/mcu/nordic/nrf52xxx/syscfg.yml]: Set default value to 0.

File: hw/mcu/nordic/nrf52xxx/syscfg.yml [hunk 6]
```diff
@@ -114,6 +114,76 @@ syscfg.defs:
...
+    QSPI_FLASH_PAGE_SIZE:
+        description: 'QSPI page size. Writes can only be perfrmed to one page at a time.'
+        value: 256
```
Comment (mkiiskila) [inline hw/mcu/nordic/nrf52xxx/syscfg.yml]: Set default value to 0.

**Pull request merged**